### PR TITLE
Pull Request from Brett

### PR DESCRIPTION
**Pull request by brettlyons**
All of the tests are set to run, and pass.  The tests are not entirely exhaustive, and the code solutions guard against unwritten test cases for the letterGrader function via deliberate type-constraint.  You could probably pass a string into one of the average functions if you wanted to break it.  I cannot, in good concience label this as finished.

### CODE REVIEW

File: grader.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    var outputString = "F";
+    var scoreMap = {
+      9: "A",
+      8: "B",
+      7: "C",
+      6: "D",
+    };
+    if ( !Number(score) || score > 100 ) {
+      return (score + " not a valid score");
+    }
```
Comment (craftninja) [inline grader.js]: Probably a good idea, but out of scope. Definitely check in with your project owner before adding code that might not be necessary or financially viable!

File: grader.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  },
+
+  //type [Number] -> String;
+  averageGrade: function ( gradesArray ) {
+    var sum = gradesArray.reduce(function ( a, b ) {
+      return a + b;
+    });
+    return String(sum / gradesArray.length);
+  },
```
Comment (craftninja) [inline grader.js]: Nice, easy to follow.

File: spec/grader_spec.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  it('letterGrader( score ) is run and returns a letter grade', function () {
+    expect(code.letterGrader(95)).toEqual("A");
+    expect(code.letterGrader(98)).toEqual("A+");
+    expect(code.letterGrader(92)).toEqual("A-");
+  });
+});
+
+describe('graderTests', function() {
+  it('letterGrader( score ) is run and returns an error msg', function () {
+    expect(code.letterGrader("lizard")).toEqual("lizard not a valid score");
```
Comment (craftninja) [inline spec/grader_spec.js]: Possibly a good idea, but definitely check with product owner first! It's hard not to think about all the edge cases and test for them, but might get you into trouble if they are unnecessary.

File: spec/grader_spec.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+describe('avgTests', function() {
+  it('medianGrade( array ) is run and returns a median score of the scores array', function () {
+    expect(code.medianGrade([52,80,80,86,94])).toEqual("80");
+  });
+});
+
+describe('avgTests', function() {
+  it('modeGrade( array ) is run and returns a mode of the scores array', function () {
+    expect(code.modeGrade([82, 92, 75, 91, 92, 89, 95, 100, 86])).toEqual("92");
```
Comment (craftninja) [inline spec/grader_spec.js]: These long arrays could be pulled out into a variable... would make the test a bit easier to read! Same could be said for long strings or any complex data structures.

File: spec/grader_spec.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+});
+
+describe('avgTests', function() {
+  it('medianGrade( array ) is run and returns a median score of the scores array', function () {
+    expect(code.medianGrade([52,80,80,86,94])).toEqual("80");
+  });
+});
+
+describe('avgTests', function() {
+  it('modeGrade( array ) is run and returns a mode of the scores array', function () {
```
Comment (craftninja) [inline spec/grader_spec.js]: Read the test descriptions in order, and attempt to make a sentence when you are putting descriptions together (`describe`s from outside in, and then the `it`). Note when you run your test (and it fails) that this "sentence" is created from the series of nested descriptions.

File: grader.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    for ( var j = 0 ; j < (gradesArray.length - countOfMax) ; j++ ) {
+      if (gradesArray[i] == gradesArray[j]) {
+        indexOfMax = j;
+        countOfMax++;
+        }
+      }
+    if (countOfMax > 1 ) { newMode = gradesArray[indexOfMax] }
+    }
+  return String(newMode);
+  }
```
Comment (craftninja) [inline grader.js]: Whitespacing for `modeGrade` is not super. Also, the code is a little more complicated than is necessary. How could you refactor this to make it more human-parsable?

**Comment by craftninja** [comment]
Watch out for some edge case scope creep... it's SUPER AWESOME that you notice edge cases, but be sure to verify that this is a concern before coding for it. Pay attention to whitespacing to help with code readability. All tests pass, Nice Job!
